### PR TITLE
Standards updates 2019 Part 1

### PR DESCRIPTION
**Pull request by losborne**
First round of Standards updates.

This PR addresses the following issues:
https://jira.mitre.org/browse/BONNIE-1998 Size/Width
https://jira.mitre.org/browse/BONNIE-1997 Quantity Conversion Operators
https://jira.mitre.org/browse/BONNIE-1996 ConvertsToTime, ToTime, ToString formatting
https://jira.mitre.org/browse/BONNIE-1995 DateTime Second+Millisecond comparison
https://jira.mitre.org/browse/BONNIE-1994 timezoneOffset Changes

In making these updates, Quantity was converted to a datatype (taking care of an existing TODO in the code).  DateTime equivalence was also implemented (it previously fell through and used equals).

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] All dependent libraries are appropriately updated or have a corresponding PR related to this change
- [x] `cql4browsers.js` built with `yarn run build-everything` if coffeescript source changed.

**Reviewer:**

Name: @cmoesel 
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: test/elm/comparison/data.coffee [hunk 1]
```diff
@@ -51,8 +53,32 @@ define EqRatios: 10 'mg' : 2 'dL' = 15 'mg' : 3 'dL'
 define UneqRatios: 10 'mg' : 2 'dL' = 15 'mg' : 4 'dL'
 ###
 
+###
+Translation Error(s):
+[11:28, 11:42] Could not resolve call to operator Equal with signature (tuple{Foo:System.Any},tuple{Bar:System.Any}).
+[15:40, 15:94] Could not resolve call to operator Equal with signature (tuple{a:System.String,b:System.Any},tuple{a:System.String,c:System.Any}).
```
Comment (cmoesel) [inline test/elm/comparison/data.coffee]: Have you looked into the errors above? Are these valid test cases that point out an issue in CQL-to-ELM? Or are they invalid test cases? If the latter, should they be fixed or removed?

File: test/elm/comparison/data.coffee [hunk 2]
```diff
@@ -3256,8 +3466,22 @@ define AEqB_Quantity_incompatible: 5 'Cel' != 5 'm'
 define ALtB_Quantity_incompatible: 5 'Cel' != 40 'm'
 ###
 
+###
+Translation Error(s):
+[12:40, 12:95] Could not resolve call to operator Equal with signature (tuple{a:System.String,b:System.Any},tuple{a:System.String,c:System.Any}).
```
Comment (cmoesel) [inline test/elm/comparison/data.coffee]: Have you looked into this error? Is this a valid test case that points out an issue in CQL-to-ELM? Or is it an invalid test case? If the latter, should it be fixed or removed?

File: test/elm/interval/data.coffee [hunk 3]
```diff
@@ -104369,8 +105845,172 @@ define HourPrecPerSecond: expand { Interval[@T01+00:00, @T01+00:00] } per second
...
+[11:44, 11:59] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[12:49, 12:64] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[14:42, 14:54] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[15:44, 15:56] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[16:44, 16:56] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[17:49, 17:61] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[19:43, 19:52] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[20:45, 20:54] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[21:45, 21:54] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
+[22:50, 22:59] Could not resolve call to operator Add with signature (System.Time,System.Ratio).
```
[the remark this answers is not in the document]
Comment (losborne) [reply test/elm/interval/data.coffee]: We have a task in the backlog to look into these errors and the ones below, but we have prioritized that investigation as lower than the standard updates.  I believe that @jbradl11 asked bryn about them and it was determined that these errors are likely things we need to update rather than issues with cql-to-elm.

File: src/datatypes/datetime.coffee [hunk 4]
```diff
@@ -334,7 +336,8 @@ class DateTime
     new Date(@year, @month, @day)
 
   getTime: () ->
-    new DateTime(0, 1, 1, @hour, @minute, @second, @millisecond, @timezoneOffset)
+    # Times no longer have timezoneOffets, so we must explicitly set it to null
+    new DateTime(0, 1, 1, @hour, @minute, @second, @millisecond, null)
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/datetime.coffee]: I agree having a separate `Time` class would be better.  We can add a backlog task for that but will not likely have time to implement along with these updates.

File: src/datatypes/datetime.coffee [hunk 5]
```diff
@@ -882,11 +866,50 @@ DateTime.prototype.getFieldCieling = Date.prototype.getFieldCieling = (field) ->
     return 999
   throw new Error('Tried to clieling a field that has no cieling value: ' + field)
 
+compareWithDefaultResult = (a, b, defaultResult) ->
+  # leave with false there is a type mismatch
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/datetime.coffee]: Updated wording

File: src/datatypes/interval.coffee [hunk 6]
```diff
@@ -291,17 +292,46 @@ module.exports.Interval = class Interval
...
+    if (@low? and (@low.isDateTime or @low.isDate or @low.isTime)) or
+       (@high? and (@high.isDateTime or @high.isDate or @high.isTime))
+      throw new Error("Width of Date, DateTime, and Time intervals is not supported")
 
     closed = @toClosed()
     if closed.low instanceof Uncertainty or closed.high instanceof Uncertainty
       null
+    else if closed.low.isQuantity
+      if closed.low.unit != closed.high.unit
+        throw new Error("Cannot calculate width of Quantity Interval with different units")
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/interval.coffee]: Nothing prohibits it but the spec does say that unit conversion is optional for quantity operators.  We do support conversion for some quantity operators, but since we are in a time crunch, we opted for this route here.

Width/Size may be simple enough to support conversion for--we can give it a try if we wrap up the other changes in time.

File: src/datatypes/interval.coffee [hunk 7]
```diff
@@ -291,17 +292,46 @@ module.exports.Interval = class Interval
...
+      highValue = closed.high.value
+      diff = Math.abs(highValue - lowValue)
+      Math.round(diff * Math.pow(10, 8)) / Math.pow(10, 8)
+      new Quantity(value: diff, unit: closed.low.unit)
     else
       # TODO: Fix precision to 8 decimals in other places that return numbers
       diff = Math.abs(closed.high - closed.low)
       Math.round(diff * Math.pow(10, 8)) / Math.pow(10, 8)
 
+  size: (pointSize) ->
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/interval.coffee]: The pointSize is calculated by `getPointSize` in `src/elm/interval.coffee`

Your comment made me realize this would be better placed in `src/datatypes/interval.coffee`, so I refactored it.

Now it is called as `interval.getPointSize()`

File: src/datatypes/quantity.coffee [hunk 8]
```diff
@@ -0,0 +1,270 @@
+{ decimalAdjust, isValidDecimal } = require '../util/math'
+ucum = require 'ucum'
+
+module.exports.Quantity = class Quantity
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/quantity.coffee]: It is mainly copy+paste.  The main differences were just exporting/requiring from src/datatypes/ instead of src/elm

File: src/elm/aggregate.coffee [hunk 9]
```diff
@@ -200,7 +200,7 @@ productValue = (list) ->
     return [null, null] if filtered.length == 0
     for item in filtered
       if item.isQuantity
-        product = Quantity.doMultiplication(product,item)
+        product = doMultiplication(product,item)
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/elm/aggregate.coffee]: I agree that would be a cleaner approach.  If JavaScript supported operator overloading, that would be even nicer.

File: src/elm/datetime.coffee [hunk 10]
```diff
@@ -8,12 +8,6 @@ module.exports.DateTime = class DateTime extends Expression
   constructor: (@json) ->
     super
 
-  # Define a simple getter to allow type-checking of this class without instanceof
-  # and in a way that survives minification (as opposed to checking constructor.name)
-  Object.defineProperties @prototype,
-    isDateTime:
-      get: -> true
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/elm/datetime.coffee]: This wasn't removed but rather it is on the src/datatype/datetime.coffee instead of both the src/elm/datetime.coffee and the src/datatype/datetime.coffee.

Since the src/elm/datetime.coffee returns the src/datatype/datetime.coffee, this was redundant.

[the remark this answers is not in the document]
Comment (losborne) [reply src/elm/datetime.coffee]: Responded too quickly before I really read your whole comment.

I don't think there was any reason to have this in both places--since the constructor here returns the datatype/datetime, the object has the `isDateTime` property already.

File: src/elm/interval.coffee [hunk 11]
```diff
@@ -121,6 +121,16 @@ module.exports.Width = class Width extends Expression
...
 
+module.exports.Size = class Size extends Expression
+  constructor: (json) ->
+    super
+
+  exec: (ctx) ->
+    interval = @arg.execute(ctx)
+    return null unless interval?
+    pointSize = getpointSize(interval)
+    interval.size(pointSize)
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/elm/interval.coffee]: Already implemented this based on your comment above before I scrolled down this far.  I agree

File: test/elm/executor/data.coffee [hunk 12]
```diff
@@ -28,8 +28,40 @@ define DEMO: Count(InDemographic w where w is true )
 define AgeSumRef : AgeSum
 ###
 
+###
+Translation Error(s):
+[13:1, 13:18] Could not resolve context name Population in model QUICK.
```
[the remark this answers is not in the document]
Comment (losborne) [reply test/elm/executor/data.coffee]: I'm not sure about this one -- I'm not very familiar with QUICK.  @jbradl11 do you know?

File: src/datatypes/datetime.coffee (hunk 4 above)
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/datetime.coffee]: Created https://jira.mitre.org/browse/BONNIE-2071 @jbradl11 for tracking

File: src/datatypes/interval.coffee (hunk 6 above)
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/interval.coffee]: Created https://jira.mitre.org/browse/BONNIE-2072 @jbradl11 for tracking

File: test/elm/executor/data.coffee (hunk 12 above)
[the remark this answers is not in the document]
Comment (jbradl11) [reply test/elm/executor/data.coffee]: The CQL needs to be updated to use the new `Unfiltered` context. This along with the other translation errors will be fixed in the next week (probably the next PR we have you review). All of the translation errors are due to our test data needing to be updated.

File: src/datatypes/quantity.coffee [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+  # Define a simple getter to allow type-checking of this class without instanceof
+  # and in a way that survives minification (as opposed to checking constructor.name)
+  Object.defineProperties @prototype,
+    isQuantity:
+      get: -> true
+
+  constructor: (json) ->
```
[the remark this answers is not in the document]
Comment (losborne) [reply src/datatypes/quantity.coffee]: I implemented this and discovered Ratio could be a datatype so added that.  Also found some opportunity for cleanup in aggregate.coffee when updating the Quantity constructors

File: src/datatypes/interval.coffee [hunk 14]
```diff
@@ -313,6 +313,7 @@ module.exports.Interval = class Interval
       Math.round(diff * Math.pow(10, 8)) / Math.pow(10, 8)
 
   size: (pointSize) ->
+    pointSize = @getPointSize()
```
Comment (cmoesel) [inline src/datatypes/interval.coffee]: You're still accepting `pointSize` as an argument above but are immediately overwriting it.  I expect that's not the intent.

File: src/elm/aggregate.coffee [hunk 15]
```diff
@@ -227,13 +234,54 @@ module.exports.AllTrue = class AllTrue extends AggregateExpression
...
+  arr.some (x) -> x.isQuantity
+
+convertAllUnits = (arr) ->
+  # convert all quantities in array to match the unit of the first item
+  converted = []
+  for quantity in arr
+    converted.push(quantity.convertUnit(arr[0].unit))
+  converted
+
+setQuantityUnit = (quantity, arr) ->
```
Comment (cmoesel) [inline src/elm/aggregate.coffee]: As best I can tell, this function isn't used anywhere, is it?

Comment (losborne) [reply src/elm/aggregate.coffee]: It's not!  It was ugly so I refactored the code to not need it anymore, then forgot to remove it.  Will remove

File: src/datatypes/interval.coffee (hunk 14 above)
Comment (losborne) [reply src/datatypes/interval.coffee]: Good catch--will remove

**Comment by cmoesel** [comment]
I've just branched `master` to a `1.3.x` branch so that after this is merged, any work related to only CQL 1.3 should go on that branch.  So... are we ready to merge this then?

**Comment by losborne** [comment]
@cmoesel Yes, ready to merge this one--we are putting the second batch of updates on branch `standards-updates-2019-part-2`, which will go into `1.3.x`

**Comment by cmoesel** [comment]
@losborne -- OK, I will merge, but I am a little confused.  This PR is `Standards updates 2019 Part 1`, and it provides some CQL 1.4 support.  But you're saying that part 2 will go into `1.3.x`?  Either way, I'm merging this one now.

**Comment by cmoesel** [comment]
*Actually*, before I merge, I want to confirm one more thing.  Lizzie asked me a while back to always use "Squash and merge" on Bonnie PRs.  This is kind of a "collection" PR though, and has good modular commits.  Do you still want me to "Squash and Merge" this one?

**Comment by jbradl11** [comment]
We already did squash merging into the feature branch. So I think this should just be merged so we can better see the updates.

**Comment by cmoesel** [comment]
Done!
